File: services/order-service/src/main/java/com/acaboumony/order/domain/entity/Order.java
package com.acaboumony.order.domain.entity;

import com.acaboumony.order.domain.enums.OrderStatus;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * JPA entity representing a customer order in the Acabou o Mony platform.
 *
 * <p>An order aggregates one or more {@link OrderItem} entries and tracks the payment
 * lifecycle from PENDING through PAID, CANCELLED, REFUNDED, etc.</p>
 */
@Entity
@Table(name = "orders")
@Getter
@Setter
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false)
    private UUID customerId;

    @Column(nullable = false)
    private UUID merchantId;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(nullable = false)
    private OrderStatus status;

    @Column(nullable = false)
    private Long totalInCents;

    private String transactionId;

    @Column(nullable = false, unique = true)
    private UUID idempotencyKey;

    private Instant expiresAt;

    @Column(nullable = false)
    private Instant createdAt;

    @Column(nullable = false)
    private Instant updatedAt;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
    private List<OrderItem> items = new ArrayList<>();

    @PrePersist
    void prePersist() {
        createdAt = updatedAt = Instant.now();
    }

    @PreUpdate
    void preUpdate() {
        updatedAt = Instant.now();
    }
}

File: src/types/order.ts
export type OrderStatus =
  | 'PENDING'
  | 'PROCESSING'
  | 'PAID'
  | 'CANCELLED'
  | 'REFUNDED'
  | 'PARTIALLY_REFUNDED';

export interface OrderItem {
  productId: string;
  description: string;
  quantity: number;
  unitPriceInCents: number;
}

export interface OrderItemDetail extends OrderItem {
  subtotalInCents: number;
}

export interface CreateOrderRequest {
  merchantId: string;
  items: OrderItem[];
}

export interface CreateOrderResponse {
  orderId: string;
  status: OrderStatus;
  totalInCents: number;
  items: OrderItemDetail[];
  expiresAt: string;
  createdAt: string;
}

export interface OrderDetail {
  orderId: string;
  customerId: string;
  merchantId: string;
  status: OrderStatus;
  totalInCents: number;
  items: OrderItemDetail[];
  transactionId?: string;
  createdAt: string;
  updatedAt: string;
  expiresAt?: string;
}

export interface OrderSummary {
  orderId: string;
  status: OrderStatus;
  totalInCents: number;
  createdAt: string;
}

export interface OrderFormItem {
  id: string;
  productId: string;
  description: string;
  quantity: number;
  unitPriceInCents: number;
}

export interface OrderFilters {
  status?: OrderStatus;
}
